webapp/src/test/java/org/opentestsystem/rdw/admin/service/impl/DefaultCsvValidationServiceTest.java

```diff
@@ -0,0 +1,233 @@
+            assertThat(failures).hasSize(1);
+            assertThat(failures.get(0).getRow()).isEqualTo(3);
+            assertThat(failures.get(0).getMessage()).contains("out-of-order");
+        }
+    }
+
+    @Test
+    public void itShouldNotValidateACsvWithAMissingGroupsPrimaryKeyRecord() throws Exception {
+        final RecordBuilder builder = recordBuilder()
+                .row(SchoolA, GroupA)
+                .row(SchoolA, EMPTY)
+                .row(SchoolB, GroupB)
+                .row(SchoolB, GroupB);
+
+        try (final CSVParser parser = builder.build()) {
+            final List<CsvValidationFailure> failures = service.validateRecords(parser.iterator(), PermissionScope.STATEWIDE);
+            assertThat(failures).hasSize(1);
+            assertThat(failures.get(0).getRow()).isEqualTo(2);
+            assertThat(failures.get(0).getMessage()).contains("group name");
+        }
+    }
+
+    @Test
+    public void itShouldNotValidateACsvWithAnUnknownOrUnauthorizedSchool() throws Exception {
+        final RecordBuilder builder = recordBuilder()
+                .row("UnknownSchool", GroupA)
+                .row(SchoolA, GroupB)
+                .row(SchoolB, GroupB)
+                .row(SchoolB, GroupB);
+
+        try (final CSVParser parser = builder.build()) {
+            final List<CsvValidationFailure> failures = service.validateRecords(parser.iterator(), PermissionScope.STATEWIDE);
+            assertThat(failures).hasSize(1);
+            assertThat(failures.get(0).getRow()).isEqualTo(1);
+            assertThat(failures.get(0).getMessage()).contains("UnknownSchool");
+        }
+    }
+
+    @Test
+    public void itShouldNotValidateACsvWithMultipleGroupSubjects() throws Exception {
+        final RecordBuilder builder = recordBuilder()
+                .row(SchoolA, GroupA)
+                .row(SchoolA, GroupB, SchoolYear, CsvSubject.ELA, null)
+                .row(SchoolA, GroupB, SchoolYear, CsvSubject.Math, null)
+                .row(SchoolB, GroupB)
+                .row(SchoolB, GroupB);
+
+        try (final CSVParser parser = builder.build()) {
+            final List<CsvValidationFailure> failures = service.validateRecords(parser.iterator(), PermissionScope.STATEWIDE);
+            assertThat(failures).hasSize(1);
+            assertThat(failures.get(0).getRow()).isEqualTo(3);
+            assertThat(failures.get(0).getMessage()).contains("[ELA]", "[Math]");
+        }
+    }
+
+    private static RecordBuilder recordBuilder() {
+        return new RecordBuilder();
+    }
+
+    private static Map<String, Integer> createHeaderMap() {
+        final List<String> orderedHeaders = newArrayList(RequiredHeaders);
+        final Map<String, Integer> headerMap = new HashMap<>();
+        for (int i=0; i < orderedHeaders.size(); i++) {
+            headerMap.put(orderedHeaders.get(i), i);
+        }
+        return headerMap;
+    }
+
+    private static class RecordBuilder {
```

Comment:
This test harness was a pita to set up.  It annoys me that the Apache CSVRecord is a "final" class with a package-local constructor.  It means that I need to actually create a CSV in order to test.